Completed onboarding challenge

leozqi opened the pull request:
# Purpose

Pull Request for [Firmware Onboarding Challenge](https://uworbital.notion.site/Firmware-Onboarding-Challenge-24340dfccb0547a5a7246ce6d7207c28).

Suggestions and tips appreciated!

# New Changes

Implemented:

- lm75bd.c readTempLM75BD function
- thermal_mgr.c thermalMgr, osHandlerLM75BD functions

# Testing

- Passed temperature sensor unit tests
- Integration tests: Over temperature detected entering 82 degC; safe-temperature resumption detected at 74 degC.

Commit:
Completed onboarding challenge

- Passed temperature sensor unit tests
- Integration tests: Over temperature detected entering 82 degC;
  safe-temperature resumption detected at 74 degC.

Implemented:

- lm75bd.c readTempLM75BD function
- thermal_mgr.c thermalMgr, osHandlerLM75BD functions

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,43 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+
+  /* Read temperature register */
+  uint8_t read_buff[CONF_READ_BUFF_SIZE] = {0};
+  errCode = i2cReceiveFrom(LM75BD_OBC_I2C_ADDR, read_buff, CONF_READ_BUFF_SIZE);
+  if (errCode != ERR_CODE_SUCCESS)
+    return errCode;
```
use `RETURN_IF_ERROR_CODE` to make sure this error gets logged as well

leozqi replied:
Used RETURN_IF_ERROR_CODE

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,43 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+  /* Write pointer register */
+  uint8_t point_buff[CONF_POINTER_BUFF_SIZE] = {0};
+  point_buff[0] = 0x00U;
+ 
+  errCode = i2cSendTo(LM75BD_OBC_I2C_ADDR, point_buff, CONF_POINTER_BUFF_SIZE);
+  if (errCode != ERR_CODE_SUCCESS)
+    return errCode;
```
use `RETURN_IF_ERROR_CODE` to make sure this error gets logged as well

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 }
 
+#define MAX_MS_WAIT 20
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
```
check if event is null

leozqi replied:
Added null checks for events

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 
+#define MAX_MS_WAIT 20
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
   /* Send an event to the thermal manager queue */
```
check if the queue has been created

leozqi replied:
Added null checks for thermalMgrQueueHandler

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+  return ERR_CODE_SUCCESS;
+}
 
+error_code_t thermalMgrSendEventISR(thermal_mgr_event_t *event) {
```
same as above

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 
+error_code_t thermalMgrSendEventISR(thermal_mgr_event_t *event) {
+  /* Send an event to the thermal manager queue */
+  xQueueSendFromISR(thermalMgrQueueHandle, (void *)event, NULL);
```
either check the return type on this or make this function void

leozqi replied:
Checked for return type and return ERR_CODE_QUEUE_FULL

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+#define MAX_MS_WAIT 20
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
   /* Send an event to the thermal manager queue */
+  xQueueSend(thermalMgrQueueHandle, (void *)event, portTICK_PERIOD_MS*MAX_MS_WAIT);
```
check if the queueSend succeeded or timed out

leozqi replied:
Modified to check for pdTRUE

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
   while (1) {
-    
+    thermal_mgr_event_t temp_evt = {0};
+    xQueueReceive(thermalMgrQueueHandle, (void *)&temp_evt, portTICK_PERIOD_MS*MAX_MS_WAIT);
```
you need to check if this queue receive succeeded or not before doing the switch statement

leozqi replied:
Yep, thanks!

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 static void thermalMgr(void *pvParameters) {
   /* Implement this task */
+  lm75bd_config_t config = *(lm75bd_config_t *) pvParameters;
+  float temp = {0};
+  error_code_t err_code = {0};
```
in general limit the scope of your variables to the smallest scope possible

leozqi replied:
Removed err_code (not necessary) and placed temp inside while block.

Navtajh04 reviewed (changes requested):
Looks good, just a couple of changes

Commit:
Use RETURN_IF_ERROR_CODE to log errors

Commit:
Check for null events and whether queue is created

- null checks for events and thermalMgrQueueHandle
- Modified thermalMgrSendEventISR to return ERR_CODE_QUEUE_FULL in event
  of xQueueSendFromISR fail
- Modified thermalMgrSendEvent to return ERR_CODE_QUEUE_FULL in event of
  xQueueSend fail

Commit:
Check if xQueueReceive succeeded before switch

Commit:
Limit scope of variables

- Correct pdTrue, pdFalse to pdTRUE, pdFALSE

leozqi commented:
Thanks for your feedback @Navtajh04 , appreciate it!

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,38 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+ */
 error_code_t readTempLM75BD(uint8_t devAddr, float *temp) {
   /* Implement this driver function */
-  
```
check if temp is NULL

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,73 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+
   while (1) {
-    
+    thermal_mgr_event_t temp_evt = {0};
```
variable names should be camelCase

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,38 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+  error_code_t errCode;
+
+  /* Write pointer register */
+  uint8_t point_buff[CONF_POINTER_BUFF_SIZE] = {0};
```
variable names should be camelCase

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,38 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+  RETURN_IF_ERROR_CODE(i2cSendTo(LM75BD_OBC_I2C_ADDR, point_buff, CONF_POINTER_BUFF_SIZE));
+
+  /* Read temperature register */
+  uint8_t read_buff[CONF_READ_BUFF_SIZE] = {0};
```
variable names should be camelCase

Navtajh04 commented on lm75bd/lm75bd.c:
```diff
@@ -25,9 +26,38 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+
+  // Concatenate D9 through D0; read data sheet sect 7.4.3
+  // <https://www.nxp.com/docs/en/data-sheet/LM75B.pdf>
+  uint16_t temp_data = ((read_buff[0] & MASK_SIGN_BIT) << 3) | (read_buff[1] >> 5);
```
variable names should be camelCase

leozqi replied:
Changed

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,73 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+    switch (temp_evt.type) {
+    case THERMAL_MGR_EVENT_MEASURE_TEMP_CMD:
+      /* Measure temperature */
+      if (readTempLM75BD(config.devAddr, &temp) == ERR_CODE_SUCCESS) {
```
log the error code if this fails

leozqi replied:
Thanks

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +41,73 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+      }
+      break;
+    case THERMAL_MGR_EVENT_OVER_TEMP_CMD:
+      if (readTempLM75BD(config.devAddr, &temp) == ERR_CODE_SUCCESS) {
```
log the error code if this fails

leozqi replied:
Thanks

Commit:
Check if temp is NULL; variables camelCase; log errors

Commit:
Log error reading temperature in thermalMgr

Commit:
Add logging header to thermal_mgr.c

Navtajh04 commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -41,20 +42,72 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+    }
+
+    float temp = {0};
+    RETURN_IF_ERROR_CODE(readTempLM75BD(config.devAddr, &temp));
```
This should just be a log if there's an error code followed by a continue since we should not be returning from FreeRTOS functions and especially since this is a void function and this would be returning an error code

Navtajh04 reviewed (changes requested):
Just 1 change other than that looks good
